### PR TITLE
Do not send dash-specific requests to masternodes before we are fully connected

### PR DESCRIPTION
**Pull request by UdjinM6**
Open all masternode connections via `CConnman::ThreadOpenMasternodeConnections` only. Queue requests and process them after some timeout.

### CODE REVIEW

File: src/masternodeman.cpp [hunk 1]
```diff
@@ -740,6 +740,48 @@ std::pair<CService, std::set<uint256> > CMasternodeMan::PopScheduledMnbRequestCo
     return std::make_pair(pairFront.first, setResult);
 }
 
+void CMasternodeMan::ProcessQueuedMnbRequests(CConnman& connman)
+{
+    std::pair<CService, std::set<uint256> > p = mnodeman.PopScheduledMnbRequestConnection();
```
Comment (codablock) [inline src/masternodeman.cpp]: Use of mnodeman is unneeded here as we are already in CMasternodeMan

File: src/privatesend-client.cpp [hunk 2]
```diff
@@ -880,31 +884,23 @@ bool CPrivateSendClient::JoinExistingQueue(CAmount nBalanceNeedsAnonymized, CCon
...
+        nSessionDenom = dsq.nDenom;
+        infoMixingMasternode = infoMn;
+
+        {
+            LOCK(cs_dsaQueue);
+            dsaQueue.push_back(std::make_pair(GetTime(), std::make_pair(infoMn.addr, CDarksendAccept(nSessionDenom, txMyCollateral))));
         }
+
+        connman.AddPendingMasternode(infoMn.addr);
+        SetState(POOL_STATE_QUEUE);
```
[the remark this answers is not in the document]
Comment (codablock) [reply src/privatesend-client.cpp]: I checked the [timeout](https://github.com/dashpay/dash/blob/19b942571366566f23fff15383dbc2fceae6709e/src/privatesend-client.cpp#L348) handling code as well and currently we'd have to wait 30 seconds until we run into a timeout if the connection attempt fails. Maybe we should also have a different timeout for the connection attempt?

File: src/privatesend-client.cpp [hunk 3]
```diff
@@ -880,31 +884,23 @@ bool CPrivateSendClient::JoinExistingQueue(CAmount nBalanceNeedsAnonymized, CCon
...
+        nTimeLastSuccessfulStep = GetTimeMillis();
+        LogPrintf("CPrivateSendClient::JoinExistingQueue -- connecting (from queue): nSessionDenom: %d (%s), addr=%s\n",
+                  nSessionDenom, CPrivateSend::GetDenominationsToString(nSessionDenom), infoMn.addr.ToString());
+        strAutoDenomResult = _("Mixing in progress...");
         return true;
     }
+    strAutoDenomResult = _("Failed to join existing mixing queues");
```
Comment (codablock) [inline src/privatesend-client.cpp]: Would something like "Failed to find mixing queue to join" better describe what went wrong? (we didn't even try to join a queue as I understand it)

File: src/masternodeman.cpp [hunk 4]
```diff
@@ -740,6 +740,48 @@ std::pair<CService, std::set<uint256> > CMasternodeMan::PopScheduledMnbRequestCo
...
+    if (!(p.first == CService() || p.second.empty())) {
+        if (connman.IsMasternodeOrDisconnectRequested(p.first)) return;
+        mnbQueue.push_back(std::make_pair(GetTime(), p));
+        connman.AddPendingMasternode(p.first);
+    }
+
+    if (!mnbQueue.empty()) {
+        // wait at least 5 seconds since we tried to open corresponding connection
```
Comment (codablock) [inline src/masternodeman.cpp]: Hmm sounds like it go wrong or at least take more time then actually needed in some cases.

What about the other way around? You always loop through all fully connected nodes and for each node that is in mnbQueue you send the messages and remove the entry from mnbQueue.
After that loop, you go through mnbQueue and remove all entries that didn't get processed for some time already (which means connection failed).
This way the good nodes are processed ASAP, while the bad nodes fall out of the queue (which is not really a queue anymore) after some time.

Same applies to ProcessQueuedMnvRequests and ProcessQueuedDsa

**Comment by UdjinM6** [comment]
Pushed few fixes. Not fixing `POOL_STATE_QUEUE`, adding TODO  instead - it works just fine as it is while fixing it requires a protobump. I'm proposing to fix this later together with some other fixes that would require a protobump.

**Pull request merged**